rust/arrow/src/datatypes.rs

```diff
@@ -151,19 +152,70 @@ make_type!(UInt64Type, u64, DataType::UInt64, 64, 0u64);
 make_type!(Float32Type, f32, DataType::Float32, 32, 0.0f32);
 make_type!(Float64Type, f64, DataType::Float64, 64, 0.0f64);
 
-/// A subtype of primitive type that represents numeric values.
-pub trait ArrowNumericType: ArrowPrimitiveType {}
-
-impl ArrowNumericType for Int8Type {}
-impl ArrowNumericType for Int16Type {}
-impl ArrowNumericType for Int32Type {}
-impl ArrowNumericType for Int64Type {}
-impl ArrowNumericType for UInt8Type {}
-impl ArrowNumericType for UInt16Type {}
-impl ArrowNumericType for UInt32Type {}
-impl ArrowNumericType for UInt64Type {}
-impl ArrowNumericType for Float32Type {}
-impl ArrowNumericType for Float64Type {}
+/// A subtype of primitive type that represents numeric values, if available a SIMD type and
+/// SIMD operations are defined in this trait.  SIMD is leveraged in the `compute` module if
+/// available.
+pub trait ArrowNumericType: ArrowPrimitiveType {
+    /// Defines the SIMD type that should be used for this numeric type
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    type Simd;
+
+    /// The number of SIMD lanes available
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn lanes() -> usize;
+
+    /// Loads a slice into a SIMD register
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn load(slice: &[Self::Native]) -> Self::Simd;
+
+    /// Performs a SIMD add operation
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn add(left: Self::Simd, right: Self::Simd) -> Self::Simd;
+
+    /// Writes a SIMD result back to a slice
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn write(simd_result: Self::Simd, slice: &mut [Self::Native]);
+}
+
+macro_rules! make_numeric_type {
+    ($impl_ty:ty, $native_ty:ty, $simd_ty:ident) => {
+        impl ArrowNumericType for $impl_ty {
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            type Simd = $simd_ty;
+
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            fn lanes() -> usize {
+                $simd_ty::lanes()
+            }
+
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            fn load(slice: &[$native_ty]) -> $simd_ty {
+                unsafe { $simd_ty::from_slice_unaligned_unchecked(slice) }
+            }
+
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            fn add(left: $simd_ty, right: $simd_ty) -> $simd_ty {
+                left + right
+            }
+
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            fn write(simd_result: $simd_ty, slice: &mut [$native_ty]) {
+                unsafe { simd_result.write_to_slice_unaligned_unchecked(slice) };
+            }
+        }
+    };
+}
+
+make_numeric_type!(Int8Type, i8, i8x64);
```

Comment:
Have you considered AVX-512? also wondering if it would be possible to support both depending on the architecture...

Reply:
I actually thought that `packed_simd` only supported up to 256, I need to update this.